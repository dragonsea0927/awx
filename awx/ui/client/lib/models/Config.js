let $log;
let Base;

function getTruncatedVersion () {
    let version;

    try {
        [version] = this.get('version').split('-');
    } catch (err) {
        $log.error(err);
    }

    return version;
}

function isOpen () {
    return this.get('license_info.license_type') === 'open';
}

function ConfigModel (method, resource, config) {
    Base.call(this, 'config', { cache: true });

    this.Constructor = ConfigModel;
    this.getTruncatedVersion = getTruncatedVersion;
    this.isOpen = isOpen;

    return this.create(method, resource, config);
}

function ConfigModelLoader (BaseModel, _$log_) {
    Base = BaseModel;
    $log = _$log_;

    return ConfigModel;
}

ConfigModelLoader.$inject = ['BaseModel', '$log'];

export default ConfigModelLoader;
