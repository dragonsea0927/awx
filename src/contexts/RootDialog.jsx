import React, { Component } from 'react';

const RootDialogContext = React.createContext({});

export class RootDialogProvider extends Component {
  constructor (props) {
    super(props);

    this.state = {
      value: {
        title: null,
        setRootDialogMessage: ({ title, bodyText, variant }) => {
          const { value } = this.state;
          this.setState({ value: { ...value, title, bodyText, variant } });
        },
        clearRootDialogMessage: () => {
          const { value } = this.state;
          this.setState({ value: { ...value, title: null, bodyText: null, variant: null } });
        },
        ...props.value,
      }
    };
  }

  render () {
    const {
      children
    } = this.props;

    const {
      value
    } = this.state;

    return (
      <RootDialogContext.Provider value={value}>
        {children}
      </RootDialogContext.Provider>
    );
  }
}

export const RootDialog = ({ children }) => (
  <RootDialogContext.Consumer>
    {value => children(value)}
  </RootDialogContext.Consumer>
);

export function withRootDialog (Child) {
  return (props) => (
    <RootDialogContext.Consumer>
      {context => <Child {...props} {...context} />}
    </RootDialogContext.Consumer>
  );
}
