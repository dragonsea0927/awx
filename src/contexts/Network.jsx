
import axios from 'axios';
import React, { Component } from 'react';

import { withRouter } from 'react-router-dom';

import { i18nMark } from '@lingui/react';

import { withRootDialog } from './RootDialog';

import APIClient from '../api';

const NetworkContext = React.createContext({});

class Provider extends Component {
  constructor (props) {
    super(props);

    this.state = {
      value: {
        api: new APIClient(axios.create({ xsrfCookieName: 'csrftoken', xsrfHeaderName: 'X-CSRFToken' })),
        handleHttpError: err => {
          if (err.response.status === 401) {
            this.handle401();
          } else if (err.response.status === 404) {
            this.handle404();
          }
          return (err.response.status === 401 || err.response.status === 404);
        },
        ...props.value
      }
    };
  }

  handle401 () {
    const { handle401, history, setRootDialogMessage } = this.props;
    if (handle401) {
      handle401();
      return;
    }
    history.replace('/login');
    setRootDialogMessage({
      bodyText: i18nMark('You have been logged out.')
    });
  }

  handle404 () {
    const { handle404, history, setRootDialogMessage } = this.props;
    if (handle404) {
      handle404();
      return;
    }
    history.replace('/home');
    setRootDialogMessage({
      title: i18nMark('404'),
      bodyText: i18nMark('Cannot find resource.'),
      variant: 'warning'
    });
  }

  render () {
    const { value } = this.state;

    const { children } = this.props;

    return (
      <NetworkContext.Provider value={value}>
        {children}
      </NetworkContext.Provider>
    );
  }
}

export { Provider as _NetworkProvider };
export const NetworkProvider = withRootDialog(withRouter(Provider));

export function withNetwork (Child) {
  return (props) => (
    <NetworkContext.Consumer>
      {context => <Child {...props} {...context} />}
    </NetworkContext.Consumer>
  );
}
