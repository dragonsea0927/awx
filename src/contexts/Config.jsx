
import React, { Component } from 'react';

import { withNetwork } from './Network';

const ConfigContext = React.createContext({});

class provider extends Component {
  constructor (props) {
    super(props);

    this.state = {
      value: {
        ansible_version: null,
        custom_virtualenvs: null,
        version: null,
        custom_logo: null,
        custom_login_info: null
      }
    };

    this.fetchConfig = this.fetchConfig.bind(this);
  }

  componentDidMount () {
    this.fetchConfig();
  }

  async fetchConfig () {
    const { api, handleHttpError } = this.props;

    try {
      const {
        data: {
          ansible_version,
          custom_virtualenvs,
          version
        }
      } = await api.getConfig();
      const {
        data: {
          custom_logo,
          custom_login_info
        }
      } = await api.getRoot();
      this.setState({
        value: {
          ansible_version,
          custom_virtualenvs,
          version,
          custom_logo,
          custom_login_info
        }
      });
    } catch (err) {
      handleHttpError(err) || this.setState({
        value: {
          ansible_version: null,
          custom_virtualenvs: null,
          version: null,
          custom_logo: null,
          custom_login_info: null
        }
      });
    }
  }

  render () {
    const {
      value: stateValue
    } = this.state;

    const { value: propsValue, children } = this.props;

    return (
      <ConfigContext.Provider value={propsValue || stateValue}>
        {children}
      </ConfigContext.Provider>
    );
  }
}

export const ConfigProvider = withNetwork(provider);

export const Config = ({ children }) => (
  <ConfigContext.Consumer>
    {value => children(value)}
  </ConfigContext.Consumer>
);
