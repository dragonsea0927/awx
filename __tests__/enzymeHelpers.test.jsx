import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { withI18n } from '@lingui/react';
import { t } from '@lingui/macro';
import { mountWithContexts, waitForElement } from './enzymeHelpers';
import { Config } from '../src/contexts/Config';
import { withNetwork } from '../src/contexts/Network';
import { withRootDialog } from '../src/contexts/RootDialog';

describe('mountWithContexts', () => {
  describe('injected I18nProvider', () => {
    test('should mount and render', () => {
      const Child = withI18n()(({ i18n }) => (
        <div>
          <span>{i18n._(t`Text content`)}</span>
        </div>
      ));
      const wrapper = mountWithContexts(<Child />);
      expect(wrapper.find('div')).toMatchSnapshot();
    });

    test('should mount and render deeply nested consumer', () => {
      const Child = withI18n()(({ i18n }) => (
        <div>{i18n._(t`Text content`)}</div>
      ));
      const Parent = () => (<Child />);
      const wrapper = mountWithContexts(<Parent />);
      expect(wrapper.find('Parent')).toMatchSnapshot();
    });
  });

  describe('injected Router', () => {
    it('should mount and render', () => {
      const wrapper = mountWithContexts(
        <div>
          <Link to="/">home</Link>
        </div>
      );

      expect(wrapper.find('div')).toMatchSnapshot();
    });

    it('should mount and render with stubbed context', () => {
      const context = {
        router: {
          history: {
            push: jest.fn(),
            replace: jest.fn(),
            createHref: jest.fn(),
          },
          route: {
            location: {},
            match: {}
          }
        }
      };
      const wrapper = mountWithContexts(
        (
          <div>
            <Link to="/">home</Link>
          </div>
        ),
        { context }
      );

      const link = wrapper.find('Link');
      expect(link).toHaveLength(1);
      link.simulate('click', { button: 0 });
      wrapper.update();

      expect(context.router.history.push).toHaveBeenCalledWith('/');
    });
  });

  describe('injected ConfigProvider', () => {
    it('should mount and render with default values', () => {
      const Foo = () => (
        <Config>
          {value => (
            <div>
              {value.custom_virtualenvs[0]}
              {value.version}
            </div>
          )}
        </Config>
      );
      const wrapper = mountWithContexts(<Foo />);
      expect(wrapper.find('Foo')).toMatchSnapshot();
    });

    it('should mount and render with custom Config value', () => {
      const config = {
        custom_virtualenvs: ['Fizz', 'Buzz'],
        version: '1.1',
      };
      const Foo = () => (
        <Config>
          {value => (
            <div>
              {value.custom_virtualenvs[0]}
              {value.version}
            </div>
          )}
        </Config>
      );
      const wrapper = mountWithContexts(
        <Foo />,
        { context: { config } }
      );
      expect(wrapper.find('Foo')).toMatchSnapshot();
    });
  });

  describe('injected Network', () => {
    it('should mount and render', () => {
      const Foo = () => (
        <div>test</div>
      );
      const Bar = withNetwork(Foo);
      const wrapper = mountWithContexts(<Bar />);
      expect(wrapper.find('Foo')).toMatchSnapshot();
    });

    it('should mount and render with stubbed api', () => {
      const network = {
        api: {
          getFoo: jest.fn().mockReturnValue('foo value'),
        },
      };
      const Foo = ({ api }) => (
        <div>{api.getFoo()}</div>
      );
      const Bar = withNetwork(Foo);
      const wrapper = mountWithContexts(<Bar />, { context: { network } });
      expect(network.api.getFoo).toHaveBeenCalledTimes(1);
      expect(wrapper.find('div').text()).toEqual('foo value');
    });
  });

  describe('injected root dialog', () => {
    it('should mount and render', () => {
      const Foo = ({ title, setRootDialogMessage }) => (
        <div>
          <span>{title}</span>
          <button
            type="button"
            onClick={() => setRootDialogMessage({ title: 'error' })}
          >
            click
          </button>
        </div>
      );
      const Bar = withRootDialog(Foo);
      const wrapper = mountWithContexts(<Bar />);

      expect(wrapper.find('span').text()).toEqual('');
      wrapper.find('button').simulate('click');
      wrapper.update();
      expect(wrapper.find('span').text()).toEqual('error');
    });

    it('should mount and render with stubbed value', () => {
      const dialog = {
        title: 'this be the title',
        setRootDialogMessage: jest.fn(),
      };
      const Foo = ({ title, setRootDialogMessage }) => (
        <div>
          <span>{title}</span>
          <button
            type="button"
            onClick={() => setRootDialogMessage('error')}
          >
            click
          </button>
        </div>
      );
      const Bar = withRootDialog(Foo);
      const wrapper = mountWithContexts(<Bar />, { context: { dialog } });

      expect(wrapper.find('span').text()).toEqual('this be the title');
      wrapper.find('button').simulate('click');
      expect(dialog.setRootDialogMessage).toHaveBeenCalledWith('error');
    });
  });

  it('should set props on wrapped component', () => {
    function TestComponent ({ text }) {
      return (<div>{text}</div>);
    }

    const wrapper = mountWithContexts(
      <TestComponent text="foo" />
    );
    expect(wrapper.find('div').text()).toEqual('foo');
    wrapper.setProps({
      text: 'bar'
    });
    expect(wrapper.find('div').text()).toEqual('bar');
  });
});

/**
 * This is a fixture for testing async components. It renders a div
 * after a short amount of time.
 */
class TestAsyncComponent extends Component {
  constructor (props) {
    super(props);
    this.state = { displayElement: false };
  }

  componentDidMount () {
    setTimeout(() => {
      this.setState({ displayElement: true });
    }, 1000);
  }

  render () {
    const { displayElement } = this.state;
    if (displayElement) {
      return (<div id="test-async-component" />);
    }
    return null;
  }
}

describe('waitForElement', () => {
  it('waits for the element and returns it', async (done) => {
    const selector = '#test-async-component';
    const wrapper = mountWithContexts(<TestAsyncComponent />);
    expect(wrapper.exists(selector)).toEqual(false);

    const elem = await waitForElement(wrapper, selector);
    expect(elem.props().id).toEqual('test-async-component');
    expect(wrapper.exists(selector)).toEqual(true);
    done();
  });

  it('eventually throws an error for elements that don\'t exist', async (done) => {
    const selector = '#does-not-exist';
    const wrapper = mountWithContexts(<div />);

    let error;
    try {
      await waitForElement(wrapper, selector);
    } catch (err) {
      error = err;
    } finally {
      expect(error).toEqual(new Error(`Element not found using ${selector}`));
      done();
    }
  });
});
