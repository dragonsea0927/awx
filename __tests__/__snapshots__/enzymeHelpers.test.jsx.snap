// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mountWithContexts injected ConfigProvider should mount and render with custom Config value 1`] = `
<Foo>
  <Config>
    <div>
      Fizz
      1.1
    </div>
  </Config>
</Foo>
`;

exports[`mountWithContexts injected ConfigProvider should mount and render with default values 1`] = `
<Foo>
  <Config>
    <div />
  </Config>
</Foo>
`;

exports[`mountWithContexts injected I18nProvider should mount and render 1`] = `
<div>
  <I18n
    update={true}
    withHash={true}
  >
    <span>
      Text content
    </span>
  </I18n>
</div>
`;

exports[`mountWithContexts injected I18nProvider should mount and render deeply nested consumer 1`] = `
<Parent>
  <Child>
    <I18n
      update={true}
      withHash={true}
    >
      <div>
        Text content
      </div>
    </I18n>
  </Child>
</Parent>
`;

exports[`mountWithContexts injected Network should mount and render 1`] = `
<Foo
  api={"/api/"}
  handleHttpError={[Function]}
>
  <div>
    test
  </div>
</Foo>
`;

exports[`mountWithContexts injected Router should mount and render 1`] = `
<div>
  <Link
    replace={false}
    to="/"
  >
    <a
      onClick={[Function]}
    >
      home
    </a>
  </Link>
</div>
`;
